### PR TITLE
 UCS/SYS: Implement sockaddr compare function

### PR DESCRIPTION
**Pull request by dmitrygladkov**
## What

Implement sockaddr compare function

## Why ?

They will be needed for #3423 and #3428 

## How ?

The same is it done for other utilities in the `ucs/sys/sock.[ch]`

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9612/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@hoopoepg @brminich

File: src/ucs/sys/sock.c [hunk 1]
```diff
@@ -237,3 +237,106 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
 
     return str;
 }
+
+ucs_status_t ucs_sockaddr_addr_cmp(const struct sockaddr *sa1,
+                                   const struct sockaddr *sa2,
+                                   int *result_p)
+{
+    if (sa1->sa_family != sa2->sa_family) {
+        ucs_error("unable to compare socket addresses with "
```
Comment (hoopoepg) [inline src/ucs/sys/sock.c]: do we really need error messages? may output to warning & return ```not equal```?

Comment (hoopoepg) [reply src/ucs/sys/sock.c]: in general it is ok for me, but my concern was about ucs_error - isn't it too aggressive? what about ucs_debug or ucs_warn?

and one more thing: what about do not report error on comparing of valid addresses even for case if family are not equal? just return "not equal"

@brminich @yosefe ideas?

File: src/ucs/sys/sock.c [hunk 2]
```diff
@@ -237,3 +237,152 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+                  "different address families: %d vs %d",
+                  sa1->sa_family, sa2->sa_family);
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_sockaddr_family_cmp(const struct sockaddr *sa1,
```
Comment (brminich) [inline src/ucs/sys/sock.c]: why do we need this func?

File: src/ucs/sys/sock.c [hunk 3]
```diff
@@ -237,3 +237,152 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
 
     return str;
 }
+
+static ucs_status_t ucs_sockaddr_cmp_check_family(const struct sockaddr *sa)
```
Comment (brminich) [inline src/ucs/sys/sock.c]: looks like this function can be moved inside `ucs_sockaddr_cmp_family_equal`.
So finally we would have just one function for check families.

File: src/ucs/sys/sock.c [hunk 4]
```diff
@@ -237,3 +237,152 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    *result_p = memcmp(&sa1->sa_family, &sa2->sa_family,
+                       sizeof(sa1->sa_family));
+    return UCS_OK;
+}
+
+ucs_status_t ucs_sockaddr_addr_cmp(const struct sockaddr *sa1,
```
Comment (brminich) [inline src/ucs/sys/sock.c]: can we just return int, like other _cmp functions do?

File: src/ucs/sys/sock.c [hunk 5]
```diff
@@ -237,3 +237,152 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+                           &UCS_SOCKET_INET6_ADDR(sa2),
+                           sizeof(UCS_SOCKET_INET6_ADDR(sa1)));
+        return UCS_OK;
+    default:
+        ucs_error("unknown address family: %d", sa1->sa_family);
+        return UCS_ERR_INVALID_PARAM;
+    }
+}
+
+ucs_status_t ucs_sockaddr_port_cmp(const struct sockaddr *sa1,
```
Comment (brminich) [inline src/ucs/sys/sock.c]: return int?

File: src/ucs/sys/sock.c [hunk 6]
```diff
@@ -237,3 +237,152 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+                           &UCS_SOCKET_INET6_PORT(sa2),
+                           sizeof(UCS_SOCKET_INET6_PORT(sa1)));
+        return UCS_OK;
+    default:
+        ucs_error("unknown address family: %d", sa1->sa_family);
+        return UCS_ERR_INVALID_PARAM;
+    }
+}
+
+ucs_status_t ucs_sockaddr_cmp(const struct sockaddr *sa1,
```
Comment (brminich) [inline src/ucs/sys/sock.c]: int?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9626/ for details (Mellanox internal link).

File: src/ucs/sys/sock.c (hunk 1 above)
Comment (yosefe) [reply src/ucs/sys/sock.c]: yes they should be not equal

File: src/ucs/sys/sock.c [hunk 7]
```diff
@@ -237,3 +237,152 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = ucs_sockaddr_cmp_check_family(sa2);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    *result_p = memcmp(&sa1->sa_family, &sa2->sa_family,
```
Comment (yosefe) [inline src/ucs/sys/sock.c]: no need to memcmp, just ==

File: src/ucs/sys/sock.h [hunk 8]
```diff
@@ -165,6 +165,95 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+ * @param [out]  result_p   Pointer to the result
+ *                          > 1 - the first byte that doesn't match in both sockaddr's
+ *                                families has a lower value in `sa1` than in `sa2`.
+ *                            0 - the contents of both sockaddr's families are equal.
+ *                          < 1 - the first byte that doesn't match in both sockaddr's
+ *                                families has a greater value in `sa1` than in `sa2`.
+ *
+ * @return UCS_OK on success or UCS_ERR_INVALID_PARAM on failure.
+ */
+ucs_status_t ucs_sockaddr_family_cmp(const struct sockaddr *sa1,
```
Comment (yosefe) [inline src/ucs/sys/sock.h]: why do we need to expose this in sock.h? it should be internal function only

File: src/ucs/sys/sock.h [hunk 9]
```diff
@@ -165,6 +165,95 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+ * @param [out]  result_p   Pointer to the result
+ *                          > 1 - the first byte that doesn't match in both sockaddr's
+ *                                ports has a lower value in `sa1` than in `sa2`.
+ *                            0 - the contents of both sockaddr's ports are equal.
+ *                          < 1 - the first byte that doesn't match in both sockaddr's
+ *                                ports has a greater value in `sa1` than in `sa2`.
+ *
+ * @return UCS_OK on success or UCS_ERR_INVALID_PARAM on failure.
+ */
+ucs_status_t ucs_sockaddr_port_cmp(const struct sockaddr *sa1,
```
Comment (yosefe) [inline src/ucs/sys/sock.h]: why do we need to expose this in sock.h? it should be internal function only

File: src/ucs/sys/sock.h [hunk 10]
```diff
@@ -165,6 +165,95 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+ * @param [out]  result_p   Pointer to the result
+ *                          > 1 - the first byte that doesn't match in both sockaddr's
+ *                                addresses has a lower value in `sa1` than in `sa2`.
+ *                            0 - the contents of both sockaddr's addresses are equal.
+ *                          < 1 - the first byte that doesn't match in both sockaddr's
+ *                                addresses has a greater value in `sa1` than in `sa2`.
+ *
+ * @return UCS_OK on success or UCS_ERR_INVALID_PARAM on failure.
+ */
+ucs_status_t ucs_sockaddr_addr_cmp(const struct sockaddr *sa1,
```
Comment (yosefe) [inline src/ucs/sys/sock.h]: why do we need to expose this in sock.h? it should be internal function only

File: src/ucs/sys/sock.h [hunk 11]
```diff
@@ -165,6 +165,95 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+ * directly to the memory block pointed to by `to` argument.
+ *
+ * @param [in/out]   to        Pointer to sockaddr structure #1 where the content
+ *                             is to be copied.
+ * @param [in]       from      Pointer to sockaddr structure #2 of the socket address
+ *                             to be copied
+ *
+ * @return UCS_OK on success or UCS_ERR_INVALID_PARAM on failure.
+ */
+ucs_status_t ucs_sockaddr_copy(struct sockaddr *to, const struct sockaddr *from);
```
Comment (yosefe) [inline src/ucs/sys/sock.h]: why do we need special wrapper for this?

File: src/ucs/sys/sock.c (hunk 3 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: this is not needed anymore

File: src/ucs/sys/sock.c (hunk 1 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: done

File: src/ucs/sys/sock.c (hunk 2 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: removed

File: src/ucs/sys/sock.c (hunk 7 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: done

File: src/ucs/sys/sock.c (hunk 4 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: done

File: src/ucs/sys/sock.c (hunk 5 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: done

File: src/ucs/sys/sock.c (hunk 6 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: done

File: src/ucs/sys/sock.h (hunk 8 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.h]: removed

File: src/ucs/sys/sock.h (hunk 10 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.h]: removed

File: src/ucs/sys/sock.h (hunk 9 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.h]: removed

File: src/ucs/sys/sock.h (hunk 11 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.h]: removed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9737/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@hoopoepg @brminich @yosefe your comments have been applied. could you take a look pls?

File: src/ucs/sys/sock.h [hunk 12]
```diff
@@ -165,6 +166,21 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
                              char *str, size_t max_size);
 
 
+/**
+ * Return a value indicating the relationships between passed sockaddr structures.
+ * 
+ * @param [in]     sa1        Pointer to sockaddr structure #1.
+ * @param [in]     sa2        Pointer to sockaddr structure #2.
+ * @param [un/out] status_p   Pointer (can be NULL) to a statrus: UCS_OK on success
```
Comment (brminich) [inline src/ucs/sys/sock.h]: `status`

File: test/gtest/ucs/test_sock.cc [hunk 13]
```diff
@@ -268,3 +268,126 @@ UCS_TEST_F(test_socket, socket_setopt) {
...
+    EXPECT_UCS_OK(status);
+
+    result1 = ucs_sockaddr_is_equal(sa1, sa2, &status);
+    EXPECT_UCS_OK(status);
+
+    // Call w/o `status` provided
+    result2 = ucs_sockaddr_is_equal(sa1, sa2, NULL);
+    EXPECT_EQ(result1, result2);
+
+    return (result1 != 0);
```
Comment (brminich) [inline test/gtest/ucs/test_sock.cc]: `return result1` ?

File: src/ucs/sys/sock.h (hunk 12 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.h]: fixed

File: test/gtest/ucs/test_sock.cc (hunk 13 above)
Comment (dmitrygladkov) [reply test/gtest/ucs/test_sock.cc]: 👍 
done

**Comment by dmitrygladkov** [comment]
@yosefe could you take a look pls?

File: src/ucs/sys/sock.c [hunk 14]
```diff
@@ -237,3 +237,60 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    status = ucs_sockaddr_check_family(sa2);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    if (sa1->sa_family != sa2->sa_family) {
```
Comment (yosefe) [inline src/ucs/sys/sock.c]: we can check this 1st, so no need to ucs_sockaddr_check_family() for both sa1 and sa2

File: src/ucs/sys/sock.c [hunk 15]
```diff
@@ -237,3 +237,60 @@ const char* ucs_sockaddr_str(const struct sockaddr *sock_addr,
...
+        goto out;
+    }
+
+    switch (sa1->sa_family) {
+    case AF_INET:
+        if ((UCS_SOCKET_INET_PORT(sa1) == UCS_SOCKET_INET_PORT(sa2)) &&
+            (memcmp(&UCS_SOCKET_INET_ADDR(sa1), &UCS_SOCKET_INET_ADDR(sa2),
+                    sizeof(UCS_SOCKET_INET_ADDR(sa1))) == 0)) {
+            result = 1;
```
Comment (yosefe) [inline src/ucs/sys/sock.c]:
```
result = ( ... == ...) && !memcmp(...) ;
break;
```

File: src/ucs/sys/sock.c (hunk 14 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: yes, we can
but it returns 0 (not equal) and sets `status=UCS_OK` if `sa1::sa_family=AF_UNIX` and `sa2::sa_family=AF_INET` socket address provided (there are tests in gtests)

done

File: src/ucs/sys/sock.c (hunk 15 above)
Comment (dmitrygladkov) [reply src/ucs/sys/sock.c]: good catch 👍 

done

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9793/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9775/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9798/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9810/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9816/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9822/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@yosefe could you review pls?

**Pull request merged**